Feature/proto remove guava

bdmogal opened the pull request:
This PR removes usages of the guava ``collect`` package. It also replaces usages of guava ``Objects`` with Java ``Objects``. Also removes usages of ``Throwables`` and ``Longs`` classes from guava.

Pending, and in a later PR:
``Joiner``, ``Splitter``, ``Preconditions``, ``CharMatcher``, ``Charsets``, ``TypeParameter``, ``TypeToken``

Jira: [CDAP-5207](https://issues.cask.co/browse/CDAP-5207)
Build: http://builds.cask.co/browse/CDAP-DUT3665

chtyim commented:
No pom file changes?

bdmogal commented:
Not yet, not completely removed. Joiner, Splitter, Preconditions, etc coming in a separate PR.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/DatasetInstanceConfiguration.java:
```diff
@@ -37,7 +36,7 @@ public String getTypeName() {
 
   public Map<String, String> getProperties() {
-    return properties == null ? ImmutableMap.<String, String>of() : properties;
+    return properties == null ? Collections.unmodifiableMap(Collections.<String, String>emptyMap()) : properties;
```
`emptyMap` is already immutable http://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#emptyMap()

anwar6953 commented on cdap-proto/src/main/java/co/cask/cdap/proto/ProgramTypes.java:
```diff
@@ -33,16 +33,16 @@
-      .put(ServiceSpecification.class, ProgramType.SERVICE)
-      .put(WorkerSpecification.class, ProgramType.WORKER)
-      .build();
+  private static final Map<Class<? extends ProgramSpecification>, ProgramType> specClassToProgramType = new HashMap<>();
```
does it matter that its no longer immutable?

chtyim replied:
It's better be `IdentityHashMap` as the key type is `Class`.

bdmogal replied:
I thought about immutability. However, seems like this one is private and never exposed as a map outside of this class, so should be ok?

anwar6953 replied:
yea, just @chtyim's comment needs addressing here then.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/QueryInfo.java:
```diff
@@ -74,6 +73,6 @@ public long getTimestamp() {
   public int compareTo(QueryInfo queryInfo) {
     // descending.
-    return Longs.compare(queryInfo.getTimestamp(), getTimestamp());
+    return ((Long) queryInfo.getTimestamp()).compareTo((getTimestamp()));
```
Instead of casting, use `Long.valueOf`.

anwar6953 commented on cdap-proto/src/main/java/co/cask/cdap/proto/TableInfo.java:
```diff
@@ -90,7 +90,7 @@ public TableInfo(String tableName, String dbName, String owner, long creationTim
     this.retention = retention;
     this.partitionKeys = partitionKeys;
-    this.parameters = ImmutableMap.copyOf(parameters);
+    this.parameters = Collections.unmodifiableMap(parameters);
```
I'm not sure if it matters - but this isn't exactly the same.
With the original code, changes to the input `parameters` would not be reflected in the unmodifiable view.

anwar6953 replied:
Same concern applies to other places you use this replacement.

chtyim replied:
Need to make a copy with `LinkedHashMap` first.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/RunRecord.java:
```diff
@@ -49,7 +49,7 @@ public RunRecord(String pid, long startTs, @Nullable Long stopTs, ProgramRunStat
     this.stopTs = stopTs;
     this.status = status;
-    this.properties = properties == null ? Maps.<String, String>newHashMap() : properties;
+    this.properties = properties == null ? new HashMap<String, String>() : properties;
```
use `Collections.emptyMap` if it is immutable.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/SystemServiceLiveInfo.java:
```diff
@@ -30,7 +29,7 @@
 
   public SystemServiceLiveInfo(List<Containers.ContainerInfo> containers) {
-    this.containers = ImmutableList.copyOf(containers);
+    this.containers = Collections.unmodifiableList(containers);
```
This is not the same. You need to make a copy too.

`Collections.unmodifiableList(new ArrayList<>(containers))`

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/TableInfo.java:
```diff
@@ -99,7 +99,7 @@ public TableInfo(String tableName, String dbName, String owner, long creationTim
     this.numBuckets = numBuckets;
     this.serde = serde;
-    this.serdeParameters = ImmutableMap.copyOf(serdeParameters);
+    this.serdeParameters = Collections.unmodifiableMap(serdeParameters);
```
Same here.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenDetail.java:
```diff
@@ -33,7 +33,7 @@
 
   public WorkflowTokenDetail(Map<String, List<NodeValueDetail>> tokenData) {
-    this.tokenData = ImmutableMap.copyOf(tokenData);
+    this.tokenData = Collections.unmodifiableMap(tokenData);
```
Copy. Actually the original code is also wrong. It only copies the map, but not the `List`.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/codec/AbstractSpecificationCodec.java:
```diff
@@ -15,9 +15,6 @@
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSet;
 import com.google.common.reflect.TypeParameter;
 import com.google.common.reflect.TypeToken;
```
Still using guava?

chtyim commented:
In general there are some misusage of `Collections.emptyXXX` (it's already immutable) and so in some places is not preserving the original immutable copyOf promise.

bdmogal commented:
@chtyim @anwar6953 I have addressed your comments. Also, I've removed all other guava usages in ``cdap-proto`` and removed the dependency as well. Please review again.

bdmogal commented:
Build is green - http://builds.cask.co/browse/CDAP-DUT3665-5

anwar6953 commented:
LGTM

chtyim commented on cdap-explore-jdbc/src/test/java/co/cask/cdap/explore/jdbc/ExplorePreparedStatementTest.java:
```diff
@@ -45,9 +45,9 @@ public void executeTest() throws Exception {
     );
 
-
+    String ns = "ns1";
```
Why there are changes in the explore module??

anwar6953 replied:
The current identifier validation allowed empty string.
This PR changed it to now allow empty strings.

bdmogal replied:
yes the current explore tests were passing empty strings as namespace. The ``CharMatcher`` from  guava used to accept them. We never caught it because you can't pass an empty string as the namespace name from REST APIs

chtyim replied:
I see. So it's bug in the platform itself. :+1:

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenDetail.java:
```diff
@@ -33,13 +34,21 @@
+  private static Map<String, List<NodeValueDetail>> deepCopy(Map<String, List<NodeValueDetail>> tokenData) {
+    Map<String, List<NodeValueDetail>> tokenDataCopy = new LinkedHashMap<>();
+    for (Map.Entry<String, List<NodeValueDetail>> entry : tokenData.entrySet()) {
+      tokenDataCopy.put(entry.getKey(), new ArrayList<>(entry.getValue()));
```
unmodifiable list for the value list.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java:
```diff
@@ -105,7 +107,10 @@ public static ApplicationId fromIdParts(Iterable<String> idString) {
+    List<String> parts = new ArrayList<>();
+    parts.add(namespace);
+    parts.add(application);
+    return Collections.unmodifiableList(parts);
```
There is a one line equivalent of this:

```java
return Collections.unmodifiableList(Arrays.asList(new String[] { namespace, application }));
```

You don't have to change it if you don't want to. Just a note.

chtyim replied:
Actually even simpler

```java
return Collections.unmodifiableList(Arrays.asList(namespace, application));
```

bdmogal replied:
yes, much better, fixed it.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/element/EntityType.java:
```diff
@@ -122,7 +122,10 @@ public static EntityType valueOfOldIdClass(Class<? extends Id> oldIdClass) {
       return (T) fromIdParts.invoke(idParts);
     } catch (Throwable t) {
-      throw Throwables.propagate(t);
+      if (t instanceof RuntimeException) {
```
There is a nicer way to write it:

```java
try {
  someMethod();
} catch (RuntimeException e) {
  throw e;
} catch (Throwable t) {
  throw new RuntimeException(t);
}
```

chtyim replied:
Always try to avoid `instanceof` and casting if possible.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/id/EntityId.java:
```diff
@@ -132,7 +120,7 @@ public final String toString() {
       result.append(part).append(IDSTRING_PART_SEPARATOR);
     }
-    if (!Iterables.isEmpty(idParts)) {
+    if (idParts.iterator().hasNext()) {
```
This can be avoided if you have the separator as the first char.

```java
StringBuilder builder = new StringBuilder(IDSTRING_TYPE_SEPARATOR);
for (String part : toIdParts()) {
  builder.append(part);
}
builder.subString(1);
```

chtyim replied:
Also, as a general rule of thumb, you can't expect two separate calls to `Iterable.iterator()` will give you the same sequence.

bdmogal replied:
:+1: Fixed

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/id/EntityId.java:
```diff
@@ -169,8 +159,10 @@ protected static String nextAndEnd(Iterator<String> iterator, String fieldName)
+    }
     StringBuilder result = new StringBuilder();
     while (iterator.hasNext()) {
       result.append(iterator.next()).append(IDSTRING_PART_SEPARATOR);
```
Same logic as above. We can avoid the if check if the string builder begins with the `IDSTRING_PART_SEPARATOR`.

chtyim commented:
Done review. Some more comments.

bdmogal commented:
@chtyim Addressed comments. Please review the last commit again. Thanks!

bdmogal commented:
@chtyim addressed comments. Could you please have another look?

chtyim commented:
LGTM

bdmogal commented:
Build succeeded - http://builds.cask.co/browse/CDAP-DUT3665-7. Squashed commits. Merging.
